src/python/grpcio/grpc/_cython/_cygrpc/grpc_eventlet.pyx.pxi

```diff
@@ -0,0 +1,691 @@
+    """Handle the timeout of a poller greenthread."""
+    try:
+        # If the waiter is no longer there, we had an unlikely race with the
+        # kick, and we don't need to signal the timeout.
+        e_poller_waiters.remove(waiter)
+        waiter.grpc_poll_timer.cancel()
+        # This will switch execution to the `e_poller_hub.switch` line in the
+        # eventlet_run_poll returning False to express we timed out.
+        waiter.switch(False)
+    except ValueError:
+        pass
+
+
+cdef void eventlet_run_poll(size_t timeout_ms) with gil:
+    """Wait for completion of any of the callbacks."""
+    global e_poller_is_set
+
+    # NOTE: When timeouts are 0 they are usually coming from the MAINLOOP, so
+    # we cannot yield, and when the poller flag is set there's no need to wait.
+    # We could check that `e_greenlet.getcurrent()` is not
+    # `e_eventlet_lib.hubs.get_hub().greenlet` to know we are not in the
+    # MAINLOOP, but that check is slower that checking `timeout_ms` is not 0.
+    if timeout_ms and not e_poller_is_set:
+        current = e_greenlet.getcurrent()
+        timer = e_LocalTimer(timeout_ms / 1000.0, e_poll_timeout, current)
+
+        # Store the timer so we can cancel it on timeout and signal
+        current.grpc_poll_timer = timer
+
+        # Add to list of polling greenthreads that kick uses to wake them up.
+        e_poller_waiters.append(current)
+        e_poller_hub.add_timer(timer)
+
+        # Now switch to the MAINLOOP.  We'll return to this greenthread once
+        # the timeout switches back with False or signal with True.  Skip
+        # resetting the poller flag if we didn't receive the signal.
+        if not e_poller_hub.switch():
+            return
+
+    e_poller_is_set = 0
+
+
+###################
+### Initializer ###
+###################
+
+cdef grpc_socket_vtable eventlet_socket_vtable
+cdef grpc_custom_resolver_vtable eventlet_resolver_vtable
+cdef grpc_custom_timer_vtable eventlet_timer_vtable
+cdef grpc_custom_poller_vtable eventlet_pollset_vtable
+
+
+def eventlet_async_callback_func(cb, args):
+    # Method used by the gRPC code to run async callback functions
+    eventlet_spawn(cb, *args)
+
+
+def init_grpc_eventlet(connection_backlog=50):
+    """Initialize Eventlet's custom IO manager.
+
+    Here we initialize the following global variables: e_eventlet_lib,
+    e_socket_lib, eventlet_spawn, e_connection_backlog, e_greenlet,
+    e_LocalTimer, e_GlobalTimer, and SOCKET_CLOSED_EXC
+    """
+    global e_eventlet_lib, e_socket_lib, eventlet_spawn, e_connection_backlog
+    global e_greenlet, e_LocalTimer, e_GlobalTimer, SOCKET_CLOSED_EXC
+
+    # Lazily import libraries
+    import eventlet
+    from eventlet.hubs import timer
+    from eventlet.support import greenlets
+
+    e_eventlet_lib = eventlet
+    SOCKET_CLOSED_EXC = IOError('Socket closed')
+    e_socket_lib = eventlet.green.socket
+    e_LocalTimer = timer.LocalTimer
+    e_GlobalTimer = timer.Timer
```

Comment:
The difference between local and global timers is that a local timer will not fire if the parent greenthread is no longer running.  To make things simpler I'll just add a `try...except` clause and cancel the timeout instead of relying on Eventlet's behavior of local timers.